src/providers/AgreementSubscription.js

```diff
@@ -0,0 +1,172 @@
+import React, { useContext, useMemo, useState, useEffect } from 'react'
+import PropTypes from 'prop-types'
+import connectAgreement from '@aragon/connect-agreement'
+import { captureErrorWithSentry } from '../sentry'
+import { createAppHook, useApp } from '@aragon/connect-react'
+import { connector } from '../current-environment'
+import { useWallet } from '../providers/Wallet'
+import { useMounted } from '../hooks/useMounted'
+
+const { agreement } = connector
+
+const useAgreement = createAppHook(connectAgreement, agreement.connectorConfig)
+const AgreementSubscriptionContext = React.createContext()
+
+function AgreementSubscriptionProvider({ children }) {
+  const { account } = useWallet()
+  const [agreementApp, agreementAppStatus] = useApp(agreement.name)
+  const [currentVersion, currentVersionStatus] = useAgreement(
+    agreementApp,
+    (app) => app.onCurrentVersion()
+  )
+  const [disputableApps, disputableAppsStatus] = useAgreement(
+    agreementApp,
+    (app) => app.onDisputableApps()
+  )
+  const [stakingFactory, stakingFactoryStatus] = useAgreement(
+    agreementApp,
+    (app) => app.stakingFactory()
+  )
+  const [signer, signerStatus] = useAgreement(
+    agreementApp,
+    (app) => (account ? app.onSigner(account) : null),
+    [account]
+  )
+  const [
+    appsWithRequirements,
+    { loading: appsWithRequirementsLoading, error: appsWithRequirementsError },
+  ] = useAppsWithRequirements(disputableApps)
+
+  // Objects as values to avoid repeated updates on every poll
+  const currentVersionUpdateValue = JSON.stringify(currentVersion)
+  const signerUpdateValue = JSON.stringify(signer)
+
+  // TODO: Tidy this once we have an improved connect-react api
+  const loading =
+    agreementAppStatus.loading ||
+    currentVersionStatus.loading ||
+    disputableAppsStatus.loading ||
+    signerStatus.loading ||
+    stakingFactoryStatus.loading ||
+    appsWithRequirementsLoading
+
+  const error =
+    agreementAppStatus.error ||
+    currentVersionStatus.error ||
+    disputableAppsStatus.error ||
+    signerStatus.error ||
+    stakingFactoryStatus.error ||
+    appsWithRequirementsError
```

Comment:
Same as previous comment, this is ugly but readable, and a side affect of having so many separate statuses.